pykechain/exceptions.py

```diff
@@ -21,24 +20,60 @@ def __init__(self, *args, **kwargs):
         :param response:
         :param kwargs:
         """
-        response = kwargs.pop('response', None)
-        self.response = response
-        self.request = kwargs.pop('request', None)
-        if (response is not None and not self.request and hasattr(response, 'request')):
+        self.response = kwargs.pop('response', None)
+
+        try:
             self.request = self.response.request
+        except AttributeError:
+            self.request = kwargs.pop('request', None)
+
         self.msg, self.traceback, self.detail = None, None, None
-        super(APIError, self).__init__(*args, **kwargs)
 
-        if isinstance(self.response, Response) and 'content' in self.response:
+        if args:
+            arg = args[0]
+            context = [arg if isinstance(arg, str) else arg.__repr__()]
+        else:
+            context = []
+
+        if self.response is not None and isinstance(self.response, Response):
+            response_json = self.response.json()
             # may be a KE-chain API response error
-            if self.response.json():
-                self.msg = self.response.json().get('msg')
-                self.traceback = self.response.json().get('traceback')
-                self.detail = self.response.json().get('detail')
+            if response_json:
+                self.msg = response_json.get('msg')
+                self.traceback = response_json.get('traceback', 'provided no traceback.')
+                self.detail = response_json.get('detail')
             else:
                 self.traceback = self.response.text
                 self.msg = self.traceback.split(r'\n'[-1])
 
+            context.extend([
+                'Server {}'.format(self.traceback),
+                'Elapsed: {}'.format(self.response.elapsed),
+            ])
+
+        if self.request is not None and isinstance(self.request, PreparedRequest):
+            context.extend([
+                'Request URL: {}'.format(self.request.url),
+                'Request method: {}'.format(self.request.method),
+            ])
+            if self.request.body:
+                import ast
+                import json
+                try:
+                    decoded_body = self.request.body.decode("UTF-8")  # Convert byte-string to string
```

Comment:
```suggestion
    decoded_body = hasattr(self.request, 'body') and self.request.body.decode("UTF-8")
                    decoded_body = self.request.body.decode("UTF-8")  # Convert byte-string to string
```

Reply:
AttributeError is on the 'decode' method